Exercise 2

Marcelloloco opened the pull request:
Added eslint with prettier, jest and Github Actions. Also restructured the directory layout of our project by moving the webapp under the root directory and added one test and two storybooks. 

Still a bunch to be done but a first feedback would be great :)

Marcelloloco committed:
Setup of eslint, jest and Github Action

Copy webapp folder into exercise 2 directory and install eslint with prettier and eslint-plugin-jest, jest and vue-test-utils
and add a file to devine Github Actions. Jest is configured to also find tests in the src directory and eslint recognises them
there as such. The configurations for eslint and jest are done in the package.json.

Marcelloloco committed:
Restructuring of project

Moving of the webapp folder under the root directory to have a consistent point of reference for
Github Actions and merges. Fix of moving into the correct directory first before executing Github Actions.

Marcelloloco committed:
Add Storybooks

Created Storybooks for News and Newsforum component and refactored Github Action file.

RespectableRuessel reviewed (commented):
The commit looks fine, but for the next time I would recommend you to separate the commit for the CI from the webapp commits (maybe even put the CI on an own branch). The background is that if you reset to an older commit you will loose the CI checks too.

TheKilljoy reviewed (commented):
I'm getting a linter error when running npm run lint. It says: "Parsing error: Unexpected closing tag "p". It may happen when the tag has already been closed by another tag. " in file  src\components\NewsForm.vue:8:10.

TheKilljoy reviewed (commented):
The storybook for "NewsForm" found one accessibility violation: "Ensures every form element has a label" for:
- input[placeholder="Title"]
- input[placeholder="Body"]

Marcelloloco committed:
Rework on NewsForm

remove unused </p> closing tag to avoid linter warning and add label to form
to improve accesability.

felixriehm reviewed (approved):
Nice and clean. Have not seen a mdx version of storybook yet. Very interesting!

medizinmensch commented on .github/workflows/Lint-Test-Status.yaml:
```diff
@@ -0,0 +1,39 @@
+        working-directory: webapp
+    strategy:
+      matrix:
+        node-version: [8.x, 10.x, 12.x]
```
You can reduce that to one version if you like

medizinmensch commented on .github/workflows/Lint-Test-Status.yaml:
```diff
@@ -0,0 +1,39 @@
+        working-directory: webapp
+    strategy:
+      matrix:
+        node-version: [8.x, 10.x, 12.x]
```
same here: 

You can reduce that to one version if you like

medizinmensch commented on .github/workflows/Lint-Test-Status.yaml:
```diff
@@ -0,0 +1,39 @@
+name: Lint-Test-Status
```
:star:   for setting up a build server that tests and lints your code

medizinmensch commented on webapp/package.json:
```diff
@@ -0,0 +1,88 @@
+    "serve": "vue-cli-service serve",
+    "build": "vue-cli-service build",
+    "test:unit": "vue-cli-service test:unit",
+    "lint": "vue-cli-service lint",
```
:star:   For adding the linter, lint fails in case of errors

medizinmensch commented on webapp/src/components/News.spec.js:
```diff
@@ -0,0 +1,16 @@
+import { shallowMount } from "@vue/test-utils";
+import News from "@/components/News.vue";
+
+describe("News.vue", () => {
```
:star:   for testing your components with jest and vue-test-utils

medizinmensch reviewed (changes requested):
Hey! 

Sorry for the late review. There are still a few things todo until Wednesday:

Remarks: 
* For the future, please use the `git mv` command to move your files instead of copy/paste them somewhere else. The latter leads to "deleting" and "creating" the files which makes the PR more difficult to read.
* Do not commit ` webapp/src/stories/` as this is not used by your project and rather generated as example data from storybook
* as pointed out by @TheKilljoy , you have a linting warning. As it's not an error, it's not a error, I won't deduct points for that. But you sould fix it
* please put your component in a dedicated folder together with it's spec and story

What's missing:

* setting up your component properties in real time with the addon-controls from stoyrbook
* you created stories for all 
* From the a11y addon, I get the error `Form elements must have labels`. 
* When   all items are deleted from the list, it shows sth. like "The list is   empty :("
* A   button on the news list toggles between descending and ascending order


Furthermore:

* :star:   for configuring jest in a way so you can locate component, story and spec in   one folder
* :star:   for creating a story for all of your components, locating them correcctly and   adding the @storybook/addon-essentials along with @storybook/addon-a11y

Questions: 
* did you do a code review already?
* did you request a code review already?

So far you got 5/10 :star:

Marcelloloco committed:
New feature in home

button added in home component to change between ascending and descending
sort order. Also a message is added which now appears when list is empty
by utilizing a v-if. Test cases implemented for the home component to check behaviour
for empty and not empty list and changing sort order.

Marcelloloco committed:
Restructoring of the project

components, tests and storybook have been moved to individual directory.

Marcelloloco committed:
Verbose parameter for tests in Github Action

Marcelloloco committed:
Deleted auto generated stries directory

Marcelloloco committed:
Change in Github Actions to only consider one version of node.

Marcelloloco committed:
Integration of addon-controls for Storybook

Marcelloloco committed:
Add Story for Home component and comment adjustment in Newsform.stories.js

Marcelloloco committed:
Addition for NewsForm

Added .spec file for NewsForm with test cases for empty title and for duplicate entry. Commit also includes fixes in NewsForm.stories, where required newsList was
added and a the path to Home was adjusted in App.vue.

Marcelloloco committed:
Code cleanup by linter

Marcelloloco commented:
Hi @medizinmensch 
thanks for the review, I changed what you asked and pushed everything. We also already requested a review and reviewed the PR from the @Systems-Development-and-Frameworks/pls-no-js guys. 
Let me know if there is still something missing.

roschaefer commented:
@RespectableRuessel @TheKilljoy :point_down: 

![Screenshot - 2020-11-09T174958 985](https://user-images.githubusercontent.com/2110676/98570875-2f6c0c80-22b4-11eb-8cbc-16810efb5ef3.png)

> - Either "Request Changes" or "Approve" *do not just "Comment"*.

This is something I learned from last winter semester: Students feel uncomfortable to choose anything else than "Comment". There is a reason why I insist on either "Approve" or "Request changes": How to proceed with the PR? Send a clear signal! Either it's ready to be merged or it must be worked on.

In real life, a "Comment" review could leave a PR somehow in limbo.

Furthermore, with "Approve" you take a part of the responsibility when you ship this code.

@felixriehm you're the person who "Approved" but you didn't fulfill this sub-task:
> - Suggest changes to the code in "Files Changed" (you should learn how to do that).

I know that many students seem to have their first experience with code reviews in this course. And that's why I make everyone discover the "Suggest changes" button :wink:

medizinmensch commented:
Okay, so last time, a few things were missing. You addressed all of them, nice!

* (done :star: ) setting up your component properties in real time with the addon-controls from storybook
* (done :star: ) you created stories for all [components]
* (done :star: ) From the a11y addon, I get the error Form elements must have labels.
* (done :star: ) When all items are deleted from the list, it shows sth. like "The list is empty :("
* (done :star: ) A button on the news list toggles between descending and ascending order

And of course you did the review for the @Systems-Development-and-Frameworks/pls-no-js guys!


That makes 10/10 :star:
